### PR TITLE
feat!: add memory usage to JSON output

### PR DESCRIPTION
**Pull request by brandonspark**
## What:
This PR adds a field to the JSON output (currently named `max_ocaml_heap_words`, which denotes the maximum number of words used by the OCaml major heap during execution of Semgrep. Notably, this output is only triggered on the  `-json_time` `semgrep-core` setting, and `--time` correspondingly for `semgrep`.

## Why:
This will be helpful, because this information can be passed up and used in our databases (particularly Metabase), so we can do things like plot memory usage against size of repository. This will help us benchmark memory, and determine how well we do when we make improvements in the future.

## How:
Added a field `max_ocaml_heap_words` to the JSON output of both Core and CLI, under the above conditions.

## Test plan:
`make test`

Observed that the `max_ocaml_heap_words` field only appears in `semgrep --time`

Closes PA-2075.

PR checklist:

- [X] Purpose of the code is [evident to future readers](https://semgrep.dev/docs/contributing/contributing-code/#explaining-code)
- [X] Tests included or PR comment includes a reproducible test plan
- [X] Documentation is up-to-date
- [X] A changelog entry was [added to changelog.d](https://semgrep.dev/docs/contributing/contributing-code/#adding-a-changelog-entry) for any user-facing change
- [X] Change has no security implications (otherwise, ping security team)

If you're unsure on any of this, please see:

- [Contribution guidelines](https://semgrep.dev/docs/contributing/contributing-code)!
- [One of the more specific guides located here](https://semgrep.dev/docs/contributing/contributing/)

### CODE REVIEW

File: semgrep-core/src/reporting/Report.ml [hunk 1]
```diff
@@ -352,7 +353,12 @@ let make_final_result results rules ~rules_parse_time =
   let extra =
     let mk_profiling () =
       let file_times = results |> Common.map get_profiling in
-      { rules; rules_parse_time; file_times }
+      {
+        rules;
+        rules_parse_time;
+        file_times;
+        max_memory_bytes = (Gc.quick_stat ()).top_heap_words;
```
Comment (emjin) [inline semgrep-core/src/reporting/Report.ml]: This is a good start. Let's leave a comment talking about some of the caveats we discussed. Most importantly, if the most amount of memory is consumed after forking, the real max needs to be multiplied by the number of processes.

Comment (emjin) [inline semgrep-core/src/reporting/Report.ml]: To get bytes multiply by `Sys.word_size`

File: cli/src/semgrep/core_runner.py [hunk 2]
```diff
@@ -871,6 +871,7 @@ def _run_rules_direct_to_semgrep_core(
 
             if ("time" in output_json) and core_output.time:
                 self._add_match_times(profiling_data, core_output.time)
+                profiling_data.set_max_memory_bytes(core_output.time.max_memory_bytes)
```
Comment (emjin) [inline cli/src/semgrep/core_runner.py]: For consistency I would rather also have this set by a `self._add_max_memory_bytes()` function

**Review by emjin — APPROVED** [review]
Left some comments. Before merging, let's also test this by doing a deep-semgrep run with `--metrics on --time` and check metabase to see that `--max-memory-bytes` shows up, then comparing that number to roughly what we get from a local run. Not going to require a test because we can't verify the number stays correct and frankly that's what I care about

File: semgrep-core/src/reporting/Report.ml (hunk 1 above)
Comment (emjin) [reply semgrep-core/src/reporting/Report.ml]: (If you don't do that change it from `max_memory_bytes` to `max_memory_words` or at the very least pretend it's also a 64-bit machine and multiply by 8)

**Comment by brandonspark** [comment]
image

it's in the raw blob being sent to metabase, though not as a column

**Pull request merged**